polkadot/xcm/xcm-executor/src/traits/token_matching.rs

```diff
@@ -101,7 +101,22 @@ impl<AssetId, Instance> MatchesNonFungibles<AssetId, Instance> for Tuple {
 		for_tuples!( #(
 			match Tuple::matches_nonfungibles(a) { o @ Ok(_) => return o, _ => () }
 		)* );
-		tracing::trace!(target: "xcm::matches_non_fungibles", asset = ?a, "did not match fungibles asset");
+		tracing::trace!(target: "xcm::matches_non_fungibles", asset = ?a, "did not match non-fungibles asset");
+		Err(Error::AssetNotHandled)
+	}
+}
+
+pub trait MatchesInstance<Id> {
+	fn matches_instance(a: &Asset) -> result::Result<Id, Error>;
```

Comment:
This is an interesting addition. Some NFTs do indeed come without a collection, like coretime regions. Do they use the same `AssetId` and an index or do they use different `AssetId`s and the `Undefined` instance, which is supposed to be used for NFTs with only one instance?

Reply:
I think it depends on the chain that provides such NFTs :)

Yet, I'd like to see the following principle. Suppose we have collection-less NFTs on a chain. In that case, in XCM, they should be identified by the same `AssetId` and varying `AssetInstance` if these NFTs are different instances of the same "type" (where "type" is defined by the operations you can do with its objects).

So, applying this principle, the combination of `AssetId` and `AssetInstance::Undefined` should identify a singleton object with associated operations unique to this particular asset.

In short, I'd say this:
* `AssetId` should identify the NFT engine (and possibly some interior "group" of NFT in that engine, like a collection)
* `AssetInstance` should identify the particular NFT instance inside the specified NFT engine.
Note: I assume the chain can host multiple NFT engines that provide different NFT-associated operations.